contracts/Market.sol

```diff
@@ -941,144 +772,72 @@ contract Market is
    * to transfer their ERC20 tokens.
    * @param amount The total purchase amount in ERC20 tokens. This is the total number of removals being
    * purchased, scaled by the price multiple.
-   * @param customFee The custom fee percentage that was paid to Nori, as an integer, specified here for
-   * inclusion in emitted events.
-   * @param customPriceMultiple The custom price that will be charged for this transaction.
+   * @param customFee The fee percentage that was paid to Nori off chain, as an integer, specified here
+   * for inclusion in emitted events.
+   * @param customPriceMultiple The price that will be charged for this transaction.
+   * @param supplier The only supplier address from which to purchase carbon removals in this transaction, or
+   * zero address if any supplier is valid.
+   * @param vintages The valid set of vintages from which to fulfill this order, empty if any vintage is valid.
    */
   function swapWithoutFeeSpecialOrder(
     address recipient,
     address purchaser,
     uint256 amount,
     uint256 customFee,
-    uint256 customPriceMultiple
+    uint256 customPriceMultiple,
+    address supplier,
+    uint256[] memory vintages
   ) external whenNotPaused onlyRole(MARKET_ADMIN_ROLE) {
     uint256 currentPrice = _priceMultiple;
     _priceMultiple = customPriceMultiple;
     _validateCertificateAmount({amount: amount});
-    (
-      uint256 countOfRemovalsAllocated,
-      uint256[] memory ids,
-      uint256[] memory amounts,
-      address[] memory suppliers
-    ) = _allocateRemovals({purchaser: purchaser, certificateAmount: amount});
-    _fulfillOrder({
-      params: FulfillOrderData({
-        chargeFee: false,
-        feePercentage: customFee,
+    uint256 countOfRemovalsAllocated;
```

Comment:
why can this be different than ids.length?

Reply:
at this point in the code `ids` is a really long array that as instantiated with as many possible removals as could be used to fulfill. it gets sliced later